### ISSUE TITLE
1.x: Rename Observable.fromAsync()

### ISSUE THREAD

**Issue by artem-zinnatullin** [ReactiveX/RxJava#4255]
`fromAsync()` sounds like it's ok to emit from multiple threads in parallel while in fact it doesn't serialize emission.

Can we rename it? For example to `Observable.fromCallback()` which also seems more obvious to use when you convert callback-styled code to `Observable`.

**Comment by akarnokd** [comment]
The suggested name is too close to `fromCallable`. Besides, the Javadoc of `AsyncEmitter` warns about sequential use:

``` java
/**
 * Abstraction over a RxJava Subscriber that allows associating
 * a resource with it and exposes the current number of downstream
 * requested amount.
 * <p>
 * The onNext, onError and onCompleted methods should be called 
 * in a sequential manner, just like the Observer's methods. The
 * other methods are threadsafe.
 *
 * @param <T> the value type to emit
 */
```

**Comment by artem-zinnatullin** [comment]
> The suggested name is too close to fromCallable.

You're worried about dynamic languages like Groovy? They're not _that_ bad!

> Besides, the Javadoc of AsyncEmitter warns about sequential use

Well, to me it looks ~like `void threadSafe()` which has note in javadoc that it's not thread safe 😁

**Comment by akarnokd** [comment]
They sound to similar and may lead to speaker embarrassment when mixed up, not to mention, users getting confused and Content assist/Intellisense bringing up both of them when typing `fromCall`.

**Comment by JakeWharton** [comment]
I don't think something being asynchronous immediately implies that it's also concurrent.

**Comment by artem-zinnatullin** [comment]
True, but not a lot of people understand that.

Only because of `AsyncTask` a lot of Android devs think that async == bg
thread.

Another point is that `fromAsync()` is probably not what developer would
look into when he/she will need to convert callback-styled api to Rx,
`fromCallback()` looks more user-friendly (in my opinion of course).

**Comment by JakeWharton** [comment]
Using an `AsyncTask` inside `fromAsync` would be perfectly fine since it's not concurrent.

`fromCallback` is clearly not usable for the reasons previously stated. I'm not against a rename, but I'm having a hard time coming up with a better name than what we already have.

**Comment by JakeWharton** [comment]
I would almost rename _both_ `fromCallable` and `fromAsync` to just overloads of `from`. We already have a few overloads of `from` for adapting various sources so I don't see why these are treated any differently than just more sources.

Additionally, the naming conventions of `fromCallable` and `fromAsync` disagree with each other and other RxJava conventions.

The first one, `fromCallable`, describes the mechanism (`Callable`) of what it's adapting. We don't see `fromIterable` or `fromFuture` so why `fromCallable`?

`fromAsync`, on the other hand, describes the behavior that it is adapting (a callback-based source). Again, we don't have `fromSynchronous` or `justSynchronous` so why `fromAsync`? Additionally, `fromAsync` is perfectly usable for non-asynchronous sources. It's merely adapting a callback-based system (the foundation of Artem's argument) but there's no reason it has to be asynchronous. It might be just be abstracting over a visitor-based API.

I would prefer both of these be renamed to `from` because that's how you adapt an external implementation of a data source into an `Observable`. It doesn't matter if it's an `Iterable`, `Future`, `Callable`, or a callback-based source. I would also rename `AsyncEmitter` to just `Emitter` with this change.

**Comment by JakeWharton** [comment]
Humorously enough, RxJava 2 actually does postfix `from` with the source (e.g., `fromArray`, `fromIterable`, etc.). I stand by my argument that both of these should be just `from()` in RxJava 1.x, and we can bikeshed the name for 2.x in a separate issue.

**Comment by akarnokd** [comment]
> I would almost rename both fromCallable and fromAsync to just overloads of from. 

I don't recommend it, Java 8 compiler gets sometimes quite confused, for example:

```
from(() -> 1)

from(() -> new Iterator() { })
```

In this case, `from()` could mean any of the prefixed operators and you'll end up casting a lot to make it right - postfix is shorter.

**Comment by JakeWharton** [comment]
Great point. In that case, I still think `fromAsync` is the wrong name based solely on the fact that its a potentially-inaccurate description of what's being adapted, but I'm not sold on the need to change it for 1.x.

**Comment by akarnokd** [comment]
Could be `fromEmitter`.

**Comment by artem-zinnatullin** [comment]
`fromEmitter` sounds better than `fromAsync` to me, 👍

**Comment by akarnokd** [comment]
See #4442.

**Comment by akarnokd** [comment]
Closing via #4442.
